table/tables/index.go

```diff
@@ -169,8 +169,22 @@ func (c *index) Create(sctx sessionctx.Context, txn kv.Transaction, indexedValue
 		// should not overwrite the key with un-commit flag.
 		// So if the key exists, just do nothing and return.
 		v, err := txn.GetMemBuffer().Get(ctx, key)
-		if err == nil && len(v) != 0 {
-			return nil, nil
+		if err == nil {
+			if len(v) != 0 {
+				return nil, nil
+			}
+			// The key is marked as deleted in the memory buffer, as the existence check is done lazily
+			// for optimistic transactions by default. The "untouched" key could still exist in the store,
+			// it's needed to commit this key to do the existence check so unset the untouched flag.
+			if !txn.IsPessimistic() {
+				keyFlags, err := txn.GetMemBuffer().GetFlags(key)
+				if err != nil {
+					return nil, err
+				}
+				if keyFlags.HasPresumeKeyNotExists() {
+					opt.Untouched = false
```

Comment:
This will recover the deleted `c0 -> r2` record, then we got the following weird case:

```sql
/* t */ delete from t where c1 is null;
-- t >> 1 rows affected
/* t */ select * from t force index(c0);
-- t >> +----+----+----+
-- t    | C0 | C1 | C2 |
-- t    +----+----+----+
-- t    +----+----+----+
/* t */ update t set c2 = 'green' where c2 between 'purple' and 'white';
-- t >> 1 rows affected
/* t */ select * from t force index(c0);
-- t >> +----+-----+-------+
-- t    | C0 | C1  |  C2   |
-- t    +----+-----+-------+
-- t    |  1 | red | green |
-- t    +----+-----+-------+
/* t */ commit;
-- t >> E1062: Duplicate entry '1' for key 'c0'
```

A risky fix is to return this function directly here since the presume-not-exist keys will be checked when committing.

Reply:
Do you mean the current fix will recover the `c0 -> r2` record?

Reply:
Yes.

Reply:
It's caused by the same situation that `unionScan` executor dose not merge unique key results, after the delete statement the newly inserted row still exists in memory buffer:
```
r1 -> 1, null, green
1 -> r1

begin;  // insert a conflict row
r2 -> 1, red, white
1 -> r2

// after delete statement.
r2 -> 1, red, white
del r1 -> xxx
del 1 -> r2

```

Reply:
> after the delete statement the newly inserted row still exists in memory buffer:

Yes, however, the index `1 -> r2` is missing.

```
// after delete statement.
r2 -> 1, red, white
del r1 -> xxx
del 1 -> r2

// after update statement.
r2 -> 1, red, white
del r1 -> xxx
1 -> r2 // it's recovered here
```

Actually, the ideal solution will be not to delete `1 -> r2` when deleting `1 -> r1`, however, it's too complex for a must-abort transaction.